packages/atclient/src/metadata.c

```diff
@@ -677,263 +681,377 @@ exit: {
-    if ((ret = atclient_atstr_append(&protocolstr, ":algo:%s", metadata.pubkeyalgo.str) != 0)) {
-      atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atstr_append failed\n");
-      goto exit;
-    }
+    sprintf(metadatastr + pos, ":algo:%s", metadata->pubkeyalgo.str);
   }
 
   if (atclient_atkey_metadata_is_encoding_initialized(metadata)) {
-    if ((ret = atclient_atstr_append(&protocolstr, ":encoding:%s", metadata.encoding.str) != 0)) {
-      atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atstr_append failed\n");
-      goto exit;
-    }
+    sprintf(metadatastr + pos, ":encoding:%s", metadata->encoding.str);
   }
 
   if (atclient_atkey_metadata_is_enckeyname_initialized(metadata)) {
-    if ((ret = atclient_atstr_append(&protocolstr, ":encKeyName:%s", metadata.enckeyname.str) != 0)) {
-      atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atstr_append failed\n");
-      goto exit;
-    }
+    sprintf(metadatastr + pos, ":encKeyName:%s", metadata->enckeyname.str);
   }
 
   if (atclient_atkey_metadata_is_encalgo_initialized(metadata)) {
-    if ((ret = atclient_atstr_append(&protocolstr, ":encAlgo:%s", metadata.encalgo.str) != 0)) {
-      atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atstr_append failed\n");
-      goto exit;
-    }
+    sprintf(metadatastr + pos, ":encAlgo:%s", metadata->encalgo.str);
   }
 
   if (atclient_atkey_metadata_is_ivnonce_initialized(metadata)) {
-    if ((ret = atclient_atstr_append(&protocolstr, ":ivNonce:%s", metadata.ivnonce.str) != 0)) {
-      atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atstr_append failed\n");
-      goto exit;
-    }
+    sprintf(metadatastr + pos, ":ivNonce:%s", metadata->ivnonce.str);
   }
 
   if (atclient_atkey_metadata_is_skeenckeyname_initialized(metadata)) {
-    if ((ret = atclient_atstr_append(&protocolstr, ":skeEncKeyName:%s", metadata.skeenckeyname.str) != 0)) {
-      atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atstr_append failed\n");
-      goto exit;
-    }
+    sprintf(metadatastr + pos, ":skeEncKeyName:%s", metadata->skeenckeyname.str);
   }
 
   if (atclient_atkey_metadata_is_skeencalgo_initialized(metadata)) {
-    if ((ret = atclient_atstr_append(&protocolstr, ":skeEncAlgo:%s", metadata.skeencalgo.str) != 0)) {
-      atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atstr_append failed\n");
-      goto exit;
-    }
+    sprintf(metadatastr + pos, ":skeEncAlgo:%s", metadata->skeencalgo.str);
   }
 
-  if (protocolstr.olen > metadatastrlen) {
-    atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "metadatastr buffer too small\n");
-    goto exit;
+  if (strlen(metadatastr) != len) {
+    atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "metadatastr length mismatch: %lu != %lu\n",
+                          strlen(metadatastr), len);
+    return 1;
   }
```

Comment:
Another form of runtime validation, we can maybe remove this once we have unit tests in place for metadata